### PR TITLE
[WIP] updates for new PHP PPA facilitating migration to another php5.6 & php7 repo

### PR DESCRIPTION
**Pull request by till**
TODO:

 - [x] config files need to be in `/etc/php/VERSION/mods-available`
 - [x] config files need to be `ext.ini` (so we can rely on the symlinks setup by the package)
 - [x] we need a `node['php-ext']['load_priortity']` per extension
 - [x] we may need to add `extension=foo.so` to ours (when we overwrite what the package does)
 - [x] review/break-up our `php.ini` template
 - [x] re-enabled phar (included in php5.6, but we need a custom config)
 - [x] re-enable zlib (included in php5.6, but we need a custom config)

DEVOPS-124

### CODE REVIEW

**Comment by till** [comment]
@gilleyj I was using: `vagrant-test/php` VM.

This is my `deploy.json`: 

```
{
    "php": {
      "ppa": {
        "package_prefix": "php5.6"
      }
    },
    "php-fpm": {
      "prefix": "",
      "exec_prefix": "/usr",
      "socketdir": "/var/run/php",
      "packages": "php5.6-fpm,php5.6-cli,php5.6-mbstring,php-memcache"
    }
}
```

That works as far as provision is concerned and php is running at least:

```
root     18342  0.0  1.0 206808 21608 ?        Ss   15:05   0:00 php-fpm: master process (/etc/php/5.6/fpm/php-fpm.conf)                      
www-data 18345  0.0  0.2 206808  4232 ?        S    15:05   0:00 php-fpm: pool www                                                            
www-data 18346  0.0  0.2 206808  4232 ?        S    15:05   0:00 php-fpm: pool www      
```

What's missing is that the configuration files (e.g. `php.ini`, `php-fpm.conf`), end up int he correct path. 

We also need to verify what is included in this PHP, but it seems the majority of what we usually install extra is already there.

**Comment by till** [comment]
Updated `deploy.json`:

```
{
    "php": {
      "ppa": {
        "package_prefix": "php5.6"
      }
    },
    "php-fpm": {
      "prefix": "",
      "exec_prefix": "/usr",
      "fpm_config": "etc/php/5.6/fpm/php.ini",
      "cli_config": "etc/php/5.6/cli/php.ini",
      "socketdir": "/var/run/php",
      "packages": "php5.6-fpm,php5.6-cli,php5.6-mbstring,php-memcache"
    }
}

**Comment by till** [comment]
@gilleyj pushed a few things, my updated `deploy.json` for `vagrant-test/php` is:

```json
{
    "php": {
      "ppa": {
        "package_prefix": "php5.6"
      }
    },
    "php-apc": {
      "package_prefix": "php"
    },
    "php-fpm": {
      "prefix": "",
      "exec_prefix": "/usr",
      "fpm_config": "etc/php/5.6/fpm/php.ini",
      "cli_config": "etc/php/5.6/cli/php.ini",
      "socketdir": "/var/run/php",
      "packages": "php5.6-fpm,php5.6-cli,php5.6-mbstring,php-memcache"
    }
}

**Comment by till** [comment]
```json
{
    "php": {
      "ppa": {
        "package_prefix": "php5.6"
      }
    },
    "php-apc": {
      "package_prefix": "php"
    },
    "php-fpm": {
      "prefix": "",
      "exec_prefix": "/usr",
      "fpm_config": "etc/php/5.6/fpm/php.ini",
      "cli_config": "etc/php/5.6/cli/php.ini",
      "pool_dir": "etc/php/5.6/fpm/pool.d",
      "socketdir": "/var/run/php",
      "pid": "/var/run/php/php5.6-fpm.pid",
      "packages": "php5.6-fpm,php5.6-cli,php5.6-mbstring,php-memcache"
    }
}

**Comment by gilleyj** [comment]
This is the web-dna I am working with.
```
{
    "php": {
      "ppa": {
        "package_prefix": "php5.6"
      }
    },
    "php-apc": {
      "package_prefix": "php"
    },
    "php-fpm": {
      "prefix": "",
      "exec_prefix": "/usr",
      "fpm_config": "etc/php/5.6/fpm/php.ini",
      "cli_config": "etc/php/5.6/cli/php.ini",
      "pool_dir": "etc/php/5.6/fpm/pool.d",
      "socketdir": "/var/run/php",
      "pid": "/var/run/php/php5.6-fpm.pid",
      "packages": "php5.6-fpm,php5.6-cli,php5.6-mbstring,php5.6-pdo-mysql,php5.6-soap,php5.6-curl,php5.6-bcmath,php5.6-dom,php5.6-tidy,php-memcache"
    }
}

**Comment by till** [comment]
Updated `web-dna.json`:

```json
{
    "php": {
      "ppa": {
        "name": "ondrejphp",
        "uri": "ppa:ondrej/php",
        "package_prefix": "php5.6"
      }
    },
    "php-apc": {
      "package_prefix": "php"
    },
    "php-fpm": {
      "prefix": "",
      "exec_prefix": "/usr",
      "fpm_config": "etc/php/5.6/fpm/php.ini",
      "cli_config": "etc/php/5.6/cli/php.ini",
      "pool_dir": "etc/php/5.6/fpm/pool.d",
      "socketdir": "/var/run/php",
      "pid": "/var/run/php/php5.6-fpm.pid",
      "packages": "php5.6-fpm,php5.6-cli,php5.6-mbstring,php5.6-pdo-mysql,php5.6-soap,php5.6-curl,php5.6-bcmath,php5.6-dom,php5.6-tidy,php-memcache"
    }
}

**Comment by gilleyj** [comment]
With a minor change to the module-xdebug I got this setup to run with the researchready vagrant no problems.  

I'm having a tremendous amount of trouble getting scholar-vagrant going however, I've engaged a couple of the scholar devs but have yet to get it to start on master let alone this branch.  I am still trying to solve.

**Comment by till** [comment]
@gilleyj I re-based this and I am working on picking 82f9e41a9a7eaf3a90f4f0b278e56f8a2e13958a to a new PR.

**Comment by till** [comment]
Also picking: f823150dd8e718b5dc869536722c75f287179e74

File: php-fpm/recipes/configure.rb [hunk 1]
```diff
@@ -52,7 +50,13 @@
   Chef::Log.info("Adding to php sendmail stmt: #{sendmail_params}")
 end
 
-template "#{etc_fpm_dir}/#{conf_fpm}" do
+execute 'update-alternatives' do
+  command '/usr/bin/update-alternatives --install /usr/sbin/php-fpm php-fpm /usr/sbin/php-fpm5.6 0'
+  creates '/usr/sbin/php-fpm'
```
Comment (fh) [inline php-fpm/recipes/configure.rb]: This implies that it only runs if `/usr/sbin/php-fpm` does not exist - wouldnt it be cleaner to update anyway?

Comment (till) [inline php-fpm/recipes/configure.rb]: @fh I pushed 2 commits addressing this and:

 1. Ensuring `php-fpm5.6`, but also `php-fpm7.0` work
 2. Wrote a test for that (assert on the `update-alternatives` command)

File: php/attributes/default.rb [hunk 2]
```diff
@@ -6,6 +6,7 @@
 default['php']['extensions']['ini_suffix'] = '-settings'
 
 default['php-apc'] = {}
+default['php-apc']['package_prefix'] = nil
```
Comment (fh) [inline php/attributes/default.rb]: @gilleyj @till - why is there a special handling for php-apc? Maybe add a quick comment so noone is confused on further refactoring later on.

**Comment by fh** [comment]
@till Looks good to me - just the one line with the missing comment could be improved. Feel free to merge after :)

**Pull request merged**